Bad method resolution on pointers

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

I tried this code:

```rust
extern "rust-intrinsic" {
    pub fn offset<T>(dst: *const T, offset: isize) -> *const T;
}

#[lang = "const_ptr"]
impl<T> *const T {
    pub const unsafe fn offset(self, count: isize) -> *const T {
        unsafe { offset(self, count) }
    }

    pub const unsafe fn add(self, count: usize) -> Self {
        unsafe { self.offset(count as isize) }
    }
}
```

I expected to see this happen: compile without error

Instead, this happened: 

```
test.rs:12:18: error: expected [* const T=T REF: 33] got [& * const T=T REF: 33]
    8 |         unsafe { offset(self, count) }
      |                         ~
......
   12 |         unsafe { self.offset(count as isize) }
      |                  ^
```

see https://godbolt.org/z/dsfG8sahP

### Meta

- What version of Rust GCC were you using, git sha if possible. b82408fd6a17762fcb6fc7b5efe9307a3e59f8b7

--- a/gcc/rust/typecheck/rust-tyty-cmp.h
+++ b/gcc/rust/typecheck/rust-tyty-cmp.h
@@ -1240,7 +1240,7 @@ public:
     : BaseCmp (base, emit_errors), base (base)
   {}
 
-  void visit (const ReferenceType &type) override
+  void visit (const PointerType &type) override
   {
     auto base_type = base->get_base ();
     auto other_base_type = type.get_base ();

--- a/gcc/testsuite/rust/compile/issue-1031.rs
+++ b/gcc/testsuite/rust/compile/issue-1031.rs
@@ -1,0 +1,16 @@
+extern "rust-intrinsic" {
+    pub fn offset<T>(dst: *const T, offset: isize) -> *const T;
+}
+
+#[lang = "const_ptr"]
+impl<T> *const T {
+    pub const unsafe fn offset(self, count: isize) -> *const T {
+        // { dg-warning "associated function is never used" "" { target *-*-* } .-1 }
+        unsafe { offset(self, count) }
+    }
+
+    pub const unsafe fn add(self, count: usize) -> Self {
+        // { dg-warning "associated function is never used" "" { target *-*-* } .-1 }
+        unsafe { self.offset(count as isize) }
+    }
+}

